File: modules/hardware-keyboard/android/src/main/java/expo/modules/hardwarekeyboard/HardwareKeyboardModule.kt
package expo.modules.hardwarekeyboard

import android.content.Context
import android.hardware.input.InputManager
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.view.InputDevice
import android.view.KeyEvent
import android.view.View
import android.view.ViewTreeObserver
import android.view.Window
import android.widget.EditText
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

// Is a physical keyboard attached to THIS device, and what is it typing? See
// the iOS twin for why this reports a keyboard and never a scanner.
//
// The second question is the one that matters at a till. A HID barcode scanner
// types its code and presses Enter, so the only way to read it is to receive
// keystrokes -- and React Native offers no global hardware-key event. The app's
// answer used to be an invisible, permanently-focused TextInput to type into,
// which worked and cost three bugs: it fought every modal for the caret, it
// swallowed everything a real keyboard typed, and on a till with "Show
// on-screen keyboard" enabled it made Android raise the soft keyboard on every
// scan, because a focused editor is an editor whatever it looks like.
//
// A key event reaches the Activity's window BEFORE any view sees it, so
// listening there needs no focus at all: nothing is focused, nothing is
// stolen, no editor exists for the IME to appear for. That is the same shape
// as the web build's `document` listener, which is why both can now share one
// state machine in JS.
// "No React tag was known" -- kept identical to the iOS twin, where an optional
// argument turned out not to be exposed to JS at all.
private const val NO_TAG = -1

class HardwareKeyboardModule : Module() {
  private val inputManager: InputManager?
    get() = appContext.reactContext?.getSystemService(Context.INPUT_SERVICE) as? InputManager

  private val listener = object : InputManager.InputDeviceListener {
    override fun onInputDeviceAdded(deviceId: Int) = emitState()
    override fun onInputDeviceRemoved(deviceId: Int) = emitState()
    override fun onInputDeviceChanged(deviceId: Int) = emitState()
  }

  // `isVirtual` is not optional here: Android's own on-screen keyboard reports
  // as an input device with SOURCE_KEYBOARD, so counting it would make every
  // phone in the world claim a hardware keyboard.
  private fun isAttached(): Boolean = InputDevice.getDeviceIds().any { id ->
    val device = InputDevice.getDevice(id) ?: return@any false
    !device.isVirtual &&
      (device.sources and InputDevice.SOURCE_KEYBOARD) == InputDevice.SOURCE_KEYBOARD &&
      device.keyboardType == InputDevice.KEYBOARD_TYPE_ALPHABETIC
  }

  private fun emitState() {
    sendEvent("onChange", mapOf("attached" to isAttached()))
  }

  // The window callback this replaced, kept so it can be put back exactly as it
  // was. Anything else in the app that wraps the callback later would be
  // discarded by a blind restore, so it is only ever restored from here.
  private var previousCallback: Window.Callback? = null

  // WHICH window was wrapped, not merely whether one was. An Activity is
  // recreated on a dev reload and on a configuration change, and its window
  // goes with it: the wrapper installed on the old one is gone, while a flag
  // saying "already wrapped" would report the new window as covered and leave
  // scanning silently dead. Observed exactly that -- a reload, and every scan
  // afterwards fell through to the view under it and pressed whatever Android
  // had focused.
  private var wrappedWindow: Window? = null

  // Whether JS still wants keys, so a returning Activity can be re-wrapped
  // without a second subscription.
  private var capturing = false

  private val main = Handler(Looper.getMainLooper())

  // A scanner sends only printable characters and a terminator, and a person
  // pressing Shift for an uppercase SKU must not break the burst. `stepWedge`
  // in JS already ignores anything whose name is longer than one character, so
  // modifiers and arrows are reported honestly rather than filtered here --
  // one less place for the two halves to disagree about what a key is.
  private fun keyName(event: KeyEvent): String? {
    when (event.keyCode) {
      KeyEvent.KEYCODE_ENTER, KeyEvent.KEYCODE_NUMPAD_ENTER -> return "Enter"
      KeyEvent.KEYCODE_TAB -> return "Tab"
    }
    // With the meta state, so a scanner sending an uppercase or symbol code
    // through Shift reports the character it actually typed.
    val unicode = event.getUnicodeChar(event.metaState)
    if (unicode == 0) return null
    return unicode.toChar().toString()
  }

  // Keys that move around the app rather than type into it. Left alone even
  // with nothing focused, so a keyboard case can still reach a field -- and so
  // Back and the volume rocker keep working, which swallowing would break in a
  // way no one would connect to a barcode scanner.
  private fun isNavigation(event: KeyEvent): Boolean = when (event.keyCode) {
    KeyEvent.KEYCODE_TAB,
    KeyEvent.KEYCODE_DPAD_UP, KeyEvent.KEYCODE_DPAD_DOWN,
    KeyEvent.KEYCODE_DPAD_LEFT, KeyEvent.KEYCODE_DPAD_RIGHT,
    KeyEvent.KEYCODE_DPAD_CENTER, KeyEvent.KEYCODE_BACK,
    KeyEvent.KEYCODE_ESCAPE, KeyEvent.KEYCODE_MENU,
    KeyEvent.KEYCODE_VOLUME_UP, KeyEvent.KEYCODE_VOLUME_DOWN -> true
    else -> false
  }

  private fun startCapture() {
    main.post {
      val window = appContext.currentActivity?.window ?: return@post
      if (wrappedWindow === window) return@post
      wrappedWindow = window
      val base = window.callback ?: return@post
      previousCallback = base
      window.callback = object : Window.Callback by base {
        override fun dispatchKeyEvent(event: KeyEvent): Boolean {
          // A field the user tapped owns its own keys, and resolves its own
          // scans through `stepFieldBurst`. Yielding to it is the one rule this
          // whole design turns on -- and it is a CHECK, evaluated here against
          // the view that actually holds focus, rather than a fight over the
          // caret. `onCheckIsTextEditor` is what the IME itself asks, so the
          // answer cannot drift from what Android believes.
          val editorFocused = window.currentFocus?.onCheckIsTextEditor() == true
          if (editorFocused) return base.dispatchKeyEvent(event)

          // Read for both halves of the pair -- the swallow below needs it --
          // but REPORTED on the down alone: a scanner's key arrives as a
          // down/up pair, and sending both would double every character.
          val name = keyName(event)
          if (event.action == KeyEvent.ACTION_DOWN && name != null) {
            // `eventTime` rather than a clock read in JS: the burst rules turn
            // on gaps of a few milliseconds, and a value stamped when the key
            // was actually delivered survives a busy bridge that a later
            // `Date.now()` would not.
            sendEvent("onKey", mapOf("key" to name, "at" to event.eventTime.toDouble()))
          }

          // With no editor focused there is nothing in this app that hardware
          // characters legitimately belong to, and letting them through is not
          // harmless: Android focuses the first focusable view to deliver them,
          // so a scan's trailing Enter presses whatever that turns out to be --
          // observed opening the photo picker from Inventory. The web half
          // prevents exactly this with `preventDefault` on the scan's keys.
          //
          // Both halves of the pair are swallowed, or the view sees an ACTION_UP
          // for a key it never received. Navigation keys are deliberately NOT
          // swallowed: Tab and the arrows are how someone with a keyboard case
          // reaches a field in the first place, and taking those would trade one
          // trapped user for another.
          return if (name != null && !isNavigation(event)) true else base.dispatchKeyEvent(event)
        }
      }
    }
  }

  private fun stopCapture() {
    main.post {
      // Restored on the window that was actually wrapped. Handing the old
      // callback to a NEW window would graft a dead Activity's chain onto a
      // live one.
      wrappedWindow?.let { window -> previousCallback?.let { window.callback = it } }
      wrappedWindow = null
      previousCallback = null
    }
  }

  // ---------------------------------------------------------------------------
  // Typing INTO the app, which is the other half of a till with a scanner
  // plugged in: Android refuses to show its own keyboard while one is attached,
  // so the app carries a keypad of its own. That keypad used to own a value and
  // write to it, which is why it served exactly one field -- the search box --
  // and left the discount, the customer search, and every form field on the
  // till untypeable.
  //
  // A keyboard does not own a value. It sends characters to whatever has focus,
  // and lets the field decide what that means. Doing the same here is what makes
  // one dock serve every input in the app without a single field knowing it
  // exists.

  // The focused editor, addressed by React tag when the caller knows it.
  //
  // `window.currentFocus` only ever answers for the ACTIVITY's window, and a
  // React Native modal is a window of its own -- so every field inside a sheet
  // (the customer search, a product form, the float count) was invisible to it,
  // which is precisely where a till most needs a keyboard. The tag comes from
  // React and does not care which window the view was mounted into.
  private fun editorFor(tag: Int): EditText? {
    val byTag = if (tag == NO_TAG) null else appContext.findView<View>(tag)
    if (byTag is EditText) return byTag
    return appContext.currentActivity?.window?.currentFocus as? EditText
  }

  private val focusedEditor: EditText?
    get() = editorFor(NO_TAG)

  // Typing must not cost the field its focus. A key goes in, the app re-renders
  // around the new value -- a customer search re-runs, a list under it changes --
  // and a field that comes out of that unfocused takes the dock down with it,
  // because the dock follows focus. Observed exactly that: the first key landed
  // and the keyboard vanished. Re-asserting focus after the edit is what keeps
  // the second key possible.
  private fun keepFocus(editor: EditText) {
    if (!editor.isFocused) editor.requestFocus()
  }

  private fun insertText(text: String, tag: Int) {
    main.post {
      val editor = editorFor(tag) ?: return@post
      keepFocus(editor)
      // Through the selection rather than appending, so the keypad respects a
      // caret the user has moved and replaces a selection they have made --
      // the things that make an editor an editor rather than a text sink.
      val start = editor.selectionStart.coerceAtLeast(0)
      val end = editor.selectionEnd.coerceAtLeast(0)
      editor.text.replace(start.coerceAtMost(end), start.coerceAtLeast(end), text)
    }
  }

  // The Enter a scanner sends after a code, from a finger instead. Without it a
  // code typed by hand can be entered but never committed, which is a real till
  // workflow -- reading a damaged barcode off the label.
  //
  // A real key event dispatched at the field, so it travels the path the
  // scanner's own terminator travels and the field's `onSubmitEditing` fires
  // exactly as it would for hardware. Sent to the view rather than the window,
  // so the capture wrapper is not involved at all.
  private fun pressEnter(tag: Int) {
    main.post {
      val editor = editorFor(tag) ?: return@post
      keepFocus(editor)
      val now = SystemClock.uptimeMillis()
      editor.dispatchKeyEvent(KeyEvent(now, now, KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_ENTER, 0))
      editor.dispatchKeyEvent(KeyEvent(now, now, KeyEvent.ACTION_UP, KeyEvent.KEYCODE_ENTER, 0))
    }
  }

  private fun deleteBackward(tag: Int) {
    main.post {
      val editor = editorFor(tag) ?: return@post
      keepFocus(editor)
      val start = editor.selectionStart.coerceAtLeast(0)
      val end = editor.selectionEnd.coerceAtLeast(0)
      if (start != end) {
        editor.text.delete(start.coerceAtMost(end), start.coerceAtLeast(end))
      } else if (start > 0) {
        editor.text.delete(start - 1, start)
      }
    }
  }

  // Focus, reported from the one place that actually knows. The dock has to
  // appear when a field is ready for typing and leave when it is not, and every
  // JS-side approximation of that -- a prop on each field, a cache read at
  // render time -- is a copy that drifts.
  private val focusListener = ViewTreeObserver.OnGlobalFocusChangeListener { _, newFocus ->
    sendEvent("onEditorFocus", mapOf("focused" to (newFocus is EditText)))
  }

  private var watchingFocus = false
  private var watchedDecor: View? = null

  private fun startFocusWatch() {
    main.post {
      val decor = appContext.currentActivity?.window?.decorView ?: return@post
      if (watchedDecor === decor) return@post
      watchedDecor = decor
      decor.viewTreeObserver.addOnGlobalFocusChangeListener(focusListener)
    }
  }

  private fun stopFocusWatch() {
    main.post {
      watchedDecor?.viewTreeObserver?.removeOnGlobalFocusChangeListener(focusListener)
      watchedDecor = null
    }
  }

  override fun definition() = ModuleDefinition {
    Name("HardwareKeyboard")

    Events("onChange", "onKey", "onEditorFocus")

    // The keypad's two verbs. Deliberately not "setValue": a keyboard that set
    // values would have to know which field it was aimed at, which is the
    // limitation this replaces.
    Function("insertText") { text: String, tag: Int -> insertText(text, tag) }
    Function("deleteBackward") { tag: Int -> deleteBackward(tag) }
    Function("pressEnter") { tag: Int -> pressEnter(tag) }
    // Answers for the first render, before any focus CHANGE has happened.
    Function("isEditorFocused") { focusedEditor != null }

    Function("isAttached") { isAttached() }

    // Absent from binaries built before key capture existed, which is exactly
    // how the JS half tells whether it may retire the invisible sink. A `true`
    // that can only come from this build is more honest than a version number
    // nobody remembers to raise.
    Function("supportsKeyEvents") { true }

    // Per event, not per module: the window is wrapped only while something is
    // actually listening for keys, so a till with scanning switched off runs
    // with the callback chain it was born with.
    OnStartObserving("onChange") { inputManager?.registerInputDeviceListener(listener, null) }
    OnStopObserving("onChange") { inputManager?.unregisterInputDeviceListener(listener) }

    OnStartObserving("onKey") { capturing = true; startCapture() }
    OnStopObserving("onKey") { capturing = false; stopCapture() }

    OnStartObserving("onEditorFocus") { watchingFocus = true; startFocusWatch() }
    OnStopObserving("onEditorFocus") { watchingFocus = false; stopFocusWatch() }

    // The Activity this app runs in is not forever: a dev reload replaces it, and
    // so does a rotation or any other configuration change. Both listeners live
    // on the Activity's window, so both have to be re-attached to whatever
    // window is in front now -- otherwise a till keeps running, looks fine, and
    // never sees another scan.
    OnActivityEntersForeground {
      if (capturing) startCapture()
      if (watchingFocus) startFocusWatch()
    }
  }
}

File: modules/hardware-keyboard/ios/HardwareKeyboardModule.swift
import ExpoModulesCore
import GameController
// For `CACurrentMediaTime`, the monotonic clock the key timestamps use.
import QuartzCore

// Is a physical keyboard attached to THIS device, and what is it typing?
//
// A HID barcode scanner is a keyboard as far as the OS is concerned, which is
// the whole reason this exists -- and also why it must never claim to have
// found a *scanner*. It has found a keyboard. Something else decides what that
// means.
//
// GCKeyboard is iOS 14+; the app's deployment target is 16.4 (ios/Podfile), so
// no availability guard is needed.
//
// `coalesced` is the Swift-projected name for Objective-C's `coalescedKeyboard`
// class property, because Swift automatically strips the redundant type-name
// suffix when importing.
//
// The second question -- what is it typing -- is answered from the same place,
// and answering it is what lets the app stop holding the keyboard hostage in an
// invisible focused text field. `keyChangedHandler` delivers hardware keys with
// nothing focused, so a scan lands without any field asking for the caret. See
// the Android twin, which reaches the same result through the window callback.
// "No React tag was known" -- see `editor(for:)`.
private let NO_TAG = -1

public class HardwareKeyboardModule: Module {
  public func definition() -> ModuleDefinition {
    Name("HardwareKeyboard")

    Events("onChange", "onKey", "onEditorFocus")

    // The keypad's two verbs -- see the Android twin for why a keyboard sends
    // characters to whatever has focus rather than owning a value. UIKit already
    // models exactly this: `UIKeyInput` is the protocol a thing implements to be
    // typed into, and every UITextField and UITextView conforms.
    Function("insertText") { (text: String, tag: Int) in
      DispatchQueue.main.async { self.editor(for: tag)?.insertText(text) }
    }

    Function("deleteBackward") { (tag: Int) in
      DispatchQueue.main.async { self.editor(for: tag)?.deleteBackward() }
    }

    // The Enter a scanner sends after a code, from a finger instead. Without it
    // a code typed by hand can be entered but never committed, which is a real
    // till workflow -- a damaged barcode read off the label.
    //
    // A newline into the field rather than a synthesised key press: UIKit gives
    // no public way to make a text field act on its return key, and React
    // Native's field treats the newline as the return for a single-line input.
    Function("pressEnter") { (tag: Int) in
      DispatchQueue.main.async { self.editor(for: tag)?.insertText("\n") }
    }

    Function("isEditorFocused") { () -> Bool in
      return Self.firstResponder() != nil
    }

    // The counterpart, and the reason the dock has an exit at all. React
    // Native's `Keyboard.dismiss()` blurs the field its own cache is holding,
    // and that cache is emptied by a field unmounting -- so a sheet closed on a
    // focused count box leaves UIKit still holding a first responder that JS
    // can no longer name, and the dock, which follows the PLATFORM's answer,
    // could never be put away. Resigning here reaches exactly that field.
    //
    // Guarded on `UIKeyInput` so this only ever lets go of a text editor: with
    // nothing focused the responder chain answers with the window, and resigning
    // that is not ours to do.
    Function("blurEditor") {
      DispatchQueue.main.async {
        guard let responder = Self.firstResponderObject(), responder is UIKeyInput else { return }
        responder.resignFirstResponder()
      }
    }

    Function("isAttached") { () -> Bool in
      return GCKeyboard.coalesced != nil
    }

    // Unlike Android's flat `true`, this asks whether keys can ACTUALLY be
    // delivered right now. GameController does not surface every keyboard --
    // notably not the Mac's in the Simulator -- and answering "yes" where no
    // key will ever arrive would retire the fallback and leave a till that
    // cannot scan at all. No keyboard also means nothing to capture, so `false`
    // there costs nothing.
    Function("supportsKeyEvents") { () -> Bool in
      return GCKeyboard.coalesced != nil
    }

    OnStartObserving("onChange") {
      NotificationCenter.default.addObserver(
        self,
        selector: #selector(self.keyboardDidConnect),
        name: .GCKeyboardDidConnect,
        object: nil
      )
      NotificationCenter.default.addObserver(
        self,
        selector: #selector(self.keyboardDidDisconnect),
        name: .GCKeyboardDidDisconnect,
        object: nil
      )
    }

    OnStopObserving("onChange") {
      NotificationCenter.default.removeObserver(self, name: .GCKeyboardDidConnect, object: nil)
      NotificationCenter.default.removeObserver(self, name: .GCKeyboardDidDisconnect, object: nil)
    }

    OnStartObserving("onEditorFocus") { self.startFocusWatch() }
    OnStopObserving("onEditorFocus") { self.stopFocusWatch() }

    OnStartObserving("onKey") {
      self.capturing = true
      self.attachKeyHandler()
      // A scanner paired after the screen opened has to start working without
      // the screen being revisited, which is why capture re-attaches on connect
      // rather than only reading whatever was there at the time.
      NotificationCenter.default.addObserver(
        self,
        selector: #selector(self.keyboardDidConnectForCapture),
        name: .GCKeyboardDidConnect,
        object: nil
      )
    }

    OnStopObserving("onKey") {
      self.capturing = false
      GCKeyboard.coalesced?.keyboardInput?.keyChangedHandler = nil
      NotificationCenter.default.removeObserver(self, name: .GCKeyboardDidConnect, object: self)
    }
  }

  private var capturing = false

  // The focused field, without reaching for anything private. `sendAction(to:
  // nil)` is UIKit's own way of addressing "whoever is first responder": the
  // action walks the responder chain and the first thing that can handle it
  // does, which is precisely the field being typed into.
  private static weak var captured: UIResponder?

  // The field to type into, addressed by React tag when the caller knows it.
  //
  // The responder chain answers for the key window only, and a React Native
  // modal is a window of its own -- so every field inside a sheet (the customer
  // search, a product form, the float count) was unreachable, which is exactly
  // where a till most needs a keyboard. The tag comes from React and does not
  // care which window the view was mounted into. Mirrors the Android twin.
  // `NO_TAG` rather than an optional argument: an `Int?` parameter is a shape
  // Expo's iOS argument conversion would not expose, and the function simply
  // did not appear on the JS side -- so the app quietly fell back to the old
  // per-screen keypad with no error anywhere. A sentinel is uglier and works on
  // both platforms.
  private func editor(for tag: Int) -> UIKeyInput? {
    if tag != NO_TAG, let view = appContext?.findView(withTag: tag, ofType: UIView.self) as? UIKeyInput {
      return view
    }
    return Self.firstResponder()
  }

  private static func firstResponder() -> UIKeyInput? {
    return firstResponderObject() as? UIKeyInput
  }

  // The responder itself rather than what it can do, because letting go of the
  // caret is a `UIResponder` verb while typing into it is a `UIKeyInput` one.
  private static func firstResponderObject() -> UIResponder? {
    captured = nil
    UIApplication.shared.sendAction(#selector(UIResponder.captureAsFirstResponder), to: nil, from: nil, for: nil)
    return captured
  }

  fileprivate static func capture(_ responder: UIResponder) {
    captured = responder
  }

  // Focus, from the one place that knows. UIKit posts these for every text
  // field and text view in the app, so the dock follows focus without a single
  // field being told it exists -- the same universality the Android twin gets
  // from the window's focus observer.
  private func startFocusWatch() {
    let center = NotificationCenter.default
    for name in [
      UITextField.textDidBeginEditingNotification,
      UITextView.textDidBeginEditingNotification,
    ] {
      center.addObserver(self, selector: #selector(editorDidFocus), name: name, object: nil)
    }
    for name in [
      UITextField.textDidEndEditingNotification,
      UITextView.textDidEndEditingNotification,
    ] {
      center.addObserver(self, selector: #selector(editorDidBlur), name: name, object: nil)
    }
  }

  private func stopFocusWatch() {
    let center = NotificationCenter.default
    for name in [
      UITextField.textDidBeginEditingNotification,
      UITextView.textDidBeginEditingNotification,
      UITextField.textDidEndEditingNotification,
      UITextView.textDidEndEditingNotification,
    ] {
      center.removeObserver(self, name: name, object: nil)
    }
  }

  @objc
  private func editorDidFocus() {
    sendEvent("onEditorFocus", ["focused": true])
  }

  @objc
  private func editorDidBlur() {
    // Read through rather than assuming: moving between two fields ends editing
    // on one and begins on the other, and the order is not guaranteed. Asking
    // who holds it now is the only answer that survives that.
    DispatchQueue.main.async {
      self.sendEvent("onEditorFocus", ["focused": Self.firstResponder() != nil])
    }
  }

  @objc
  private func keyboardDidConnect() {
    sendEvent("onChange", ["attached": true])
  }

  @objc
  private func keyboardDidDisconnect() {
    // Read through rather than assuming false: a device can have two keyboards
    // attached, and one going away does not mean none are left.
    sendEvent("onChange", ["attached": GCKeyboard.coalesced != nil])
  }

  @objc
  private func keyboardDidConnectForCapture() {
    guard capturing else { return }
    attachKeyHandler()
  }

  private func attachKeyHandler() {
    guard let input = GCKeyboard.coalesced?.keyboardInput else { return }
    input.keyChangedHandler = { [weak self] keyboard, _, keyCode, pressed in
      // Presses only. A scanner sends each character as a down/up pair, and
      // reporting both would double every character in the buffer.
      guard pressed, let self else { return }
      let shifted = keyboard.button(forKeyCode: .leftShift)?.isPressed == true
        || keyboard.button(forKeyCode: .rightShift)?.isPressed == true
      guard let key = Self.keyName(keyCode, shifted: shifted) else { return }
      // Monotonic, like Android's `eventTime`: the burst rules turn on gaps of
      // a few milliseconds, and a wall clock can step sideways mid-scan.
      let at = CACurrentMediaTime() * 1000
      self.sendEvent("onKey", ["key": key, "at": at])
    }
  }

  // A `KeyboardEvent.key` value, the DOM's vocabulary -- a single character for
  // printable keys, or a name for the rest -- because the JS half feeds one
  // state machine shared with the web build.
  //
  // Only what a scanner can send. Every symbology in retail use is drawn from
  // digits, letters, and a handful of punctuation; a key outside this table
  // returns nil and the burst simply does not see it, which is the same
  // treatment `stepWedge` gives modifiers and arrows anyway.
  private static func keyName(_ code: GCKeyCode, shifted: Bool) -> String? {
    if code == .returnOrEnter || code == .keypadEnter { return "Enter" }
    if code == .tab { return "Tab" }
    if code == .spacebar { return " " }

    let letters: [GCKeyCode: String] = [
      .keyA: "a", .keyB: "b", .keyC: "c", .keyD: "d", .keyE: "e", .keyF: "f",
      .keyG: "g", .keyH: "h", .keyI: "i", .keyJ: "j", .keyK: "k", .keyL: "l",
      .keyM: "m", .keyN: "n", .keyO: "o", .keyP: "p", .keyQ: "q", .keyR: "r",
      .keyS: "s", .keyT: "t", .keyU: "u", .keyV: "v", .keyW: "w", .keyX: "x",
      .keyY: "y", .keyZ: "z",
    ]
    if let letter = letters[code] {
      return shifted ? letter.uppercased() : letter
    }

    // Unshifted digits and their shifted symbols, as a US layout sends them.
    // A scanner is a keyboard emulating that layout unless someone has
    // configured it otherwise, which is a shop-specific problem no table here
    // can answer.
    let digits: [GCKeyCode: (String, String)] = [
      .one: ("1", "!"), .two: ("2", "@"), .three: ("3", "#"), .four: ("4", "$"),
      .five: ("5", "%"), .six: ("6", "^"), .seven: ("7", "&"), .eight: ("8", "*"),
      .nine: ("9", "("), .zero: ("0", ")"),
    ]
    if let pair = digits[code] {
      return shifted ? pair.1 : pair.0
    }

    let keypad: [GCKeyCode: String] = [
      .keypad1: "1", .keypad2: "2", .keypad3: "3", .keypad4: "4", .keypad5: "5",
      .keypad6: "6", .keypad7: "7", .keypad8: "8", .keypad9: "9", .keypad0: "0",
    ]
    if let digit = keypad[code] { return digit }

    let punctuation: [GCKeyCode: (String, String)] = [
      .hyphen: ("-", "_"), .equalSign: ("=", "+"), .period: (".", ">"),
      .comma: (",", "<"), .slash: ("/", "?"), .semicolon: (";", ":"),
      .quote: ("'", "\""), .backslash: ("\\", "|"),
      .openBracket: ("[", "{"), .closeBracket: ("]", "}"),
    ]
    if let pair = punctuation[code] {
      return shifted ? pair.1 : pair.0
    }

    return nil
  }
}

// The selector `firstResponder()` sends into the responder chain. An extension
// on UIResponder rather than a subclass, because the thing that answers is
// whatever UIKit already put there -- a UITextField inside React Native's own
// view, which this file does not own and must not replace.
private extension UIResponder {
  @objc func captureAsFirstResponder() {
    HardwareKeyboardModule.capture(self)
  }
}
